plutus-tx/src/Language/PlutusTx/Lift/Class.hs

```diff
@@ -357,28 +358,28 @@ compileConstructorDecl TH.ConstructorInfo{TH.constructorName=name, TH.constructo
               pure $ VarDecl () constrName constrTy
           ||]
 
-makeTypeable :: TH.Type -> TH.Name -> TH.Q [TH.Dec]
-makeTypeable uni name = do
+makeTypeable :: TH.Type -> TH.Type -> TH.Name -> TH.Q [TH.Dec]
+makeTypeable uni fun name = do
     requireExtension TH.ScopedTypeVariables
 
     info <- TH.reifyDatatype name
     (rhs, deps) <- flip runReaderT mempty $ flip runStateT mempty $ (compileTypeRep info)
 
     -- See Note [Closed constraints]
-    let constraints = filter (not . isClosedConstraint) $ toConstraint uni <$> Set.toList deps
+    let constraints = filter (not . isClosedConstraint) $ toConstraint uni fun <$> Set.toList deps
 
     decl <- TH.funD 'typeRep [TH.clause [TH.wildP] (TH.normalB (TH.unTypeQ rhs)) []]
-    pure [TH.InstanceD Nothing constraints (typeablePir uni (TH.ConT name)) [decl]]
+    pure [TH.InstanceD Nothing constraints (typeablePir uni fun (TH.ConT name)) [decl]]
 
 -------
 -- Lift
 -------
 
 -- | Class for types which can be lifted into Plutus IR. Instances should be derived, do not write your
 -- own instance!
-class Lift uni a where
+class Lift uni fun a where
     -- | Get a Plutus IR term corresponding to the given value.
-    lift :: a -> RTCompile uni (Term TyName Name uni ())
+    lift :: a -> RTCompile uni fun (Term TyName Name uni fun ())
```

Comment:
TODO: try not having `fun` in `Lift uni fun a`? I think we do need it if we are going to ever have polymorphic built-in types, but perhaps not now. But having `fun` there doesn't hurt. @michaelpj WDYT?

Reply:
I *believe* that nothing in `Lift` cares about builtin functions. `Lift` only creates "data", there are never going to be any invocations of builtin functions. The main reason that might not be true is if we maybe used some builtin functions to deal with literals or something? But I don't think we do.

Reply:
> Lift only creates "data", there are never going to be any invocations of builtin functions.

So far my believe is that the only sensible way of handling dynamic built-in types is by constructing them with polymorphic built-in functions (regardless of whether it's `Map.fromList` or `List.(:)`). You can't just embed whatever you want into `Constant`, because diving into constants to see if they have some PLC terms in them is terrible: it's easy to forget that, but it'd be required by renaming, type checking, variable usage analysis, you name it.

So it makes sense not to allow for construction of terms with PLC terms in constants. Instead you put all the terms as arguments to built-in functions and they elaborate to constants during evaluation.

But we shouldn't need this now, I agree. I'll try changing this part of code and see if it works out.

Reply:
Done. There's a comment missing, though.

Reply:
Added the comment.